### PR TITLE
Minor updates to transmon docs

### PR DESCRIPTION
**Pull request by BorjaRequena**
Working with the pulse module, I spotted a math typo in the transmon drive docstring. I have taken the chance to add a few changes that I believe would have made my first contact with the docs a bit smoother.

### CODE REVIEW

File: pennylane/pulse/transmon.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def transmon_interaction(
     see e.g. `arXiv:1804.04073 <https://arxiv.org/abs/1804.04073>`_,
     `arXiv:2203.06818 <https://arxiv.org/abs/2203.06818>`_, or `arXiv:2210.15812 <https://arxiv.org/abs/2210.15812>`_.
 
-    .. note:: Currently only supporting ``d=2`` with qudit support planned in the future.
+    .. note:: Currently only supporting ``d=2`` with qudit support planned in the future. For ``d=2``, we have :math:`a:=\frac{1}{2}(\sigma^x + i \sigma^y)`.
```
Comment (eddddddy) [inline pennylane/pulse/transmon.py:75]: I know you're just copying the other docstring, but why is this `sigma^x` instead of `sigma_x`? In other places in the docs we write `sigma_x` for the Pauli X matrix.

File: pennylane/pulse/transmon.py [hunk 2]
```diff
@@ -299,22 +299,24 @@ def transmon_drive(amplitude, phase, freq, wires, d=2):
...
+        H = qml.pulse.transmon_drive(amp, phase, freq, 0)
 
         t = 0.5
         A = 0.1
         phi = 0.001
         params = [A, phi]
 
-    Evaluated at :math:`t = \frac{1}{2}` with the parameters :math:`A = 0.1` and :math:`\phi = 10^{-3}` we obtain
-    :math:`2 \pi A \left(\frac{1}{2}(\sigma^x + i \sigma^y) + \frac{1}{2}(\sigma^x + i \sigma^y)\right) = 2 \pi A \sigma^x = 0.63 \sigma^x`.
+    Evaluated at :math:`t = \frac{1}{2}` with the parameters :math:`A = 0.1` and :math:`\phi_0 = 10^{-3}` we obtain
```
Comment (mudit2812) [inline pennylane/pulse/transmon.py:318]:
```suggestion
    Evaluated at :math:`t = \frac{1}{2}` with the parameters :math:`A = 0.1` and :math:`\phi = 10^{-3}` we obtain
```
The phase variable is called `phi` on line 313/315

File: pennylane/pulse/transmon.py [hunk 3]
```diff
@@ -338,7 +340,7 @@ def phase(phi0, t): return phi0 * (t - 0.5)
         H = qml.pulse.transmon_interaction(qubit_freqs, connections, g, wires=range(3))
 
         def amp(max_amp, t): return max_amp * jnp.sin(t) ** 2
-        def freq(fr, t): return fr
+        def freq(fr, t): return fr  # Enables frequency trainability
```
Comment (mudit2812) [inline pennylane/pulse/transmon.py:343]: I'm not a fan of the way frequency is defined here. Is casting a constant frequency is a tensor not enough? @lillian542 @dwierichs

**Review by mudit2812 — COMMENTED** [review]
One comment that I want to get some more clarity about, but otherwise good to go. I'll approve once the comment is addressed.

File: pennylane/pulse/transmon.py (hunk 1 above)
Comment (eddddddy) [reply pennylane/pulse/transmon.py:75]: Oh I didn't know the sub/superscripts had this much discrepancy across the docs. Sure, you've convinced me. Maybe we can just stick the tidying up on the backlog then.

File: pennylane/pulse/transmon.py (hunk 2 above)
Comment (BorjaRequena) [reply pennylane/pulse/transmon.py:318]: Yep. However, in the math we have $\phi(t) = \phi_0(t - 1/2)$ and the variable `phi` actually denotes $\phi_0$. We could rename the variable to `phi_0` to make it explicit. In this part of the example, since we take $t=1/2$, $\phi_0$ is irrelevant actually.

File: pennylane/pulse/transmon.py (hunk 3 above)
Comment (BorjaRequena) [reply pennylane/pulse/transmon.py:343]: The frequency is handled this way in order to enable the trainability of the parameter. It is mentioned in the next paragraph so, in a first read, it does not make much sense. This is one of the things that got me wondering, which is why I added the comment. Maybe we could even add "see below" in the comment.

Comment (dwierichs) [reply pennylane/pulse/transmon.py]: We have the function `qml.pulse.constant` which exists solely for this purpose. In my opinion it would be nice to use that function here mainly for two reasons:
1. It makes clear that just passing a number is not the same, because "otherwise the PL devs would not have written a special function for this".
2. It will encourage users to do so as well.

Comment (mudit2812) [reply pennylane/pulse/transmon.py]: Thanks both of you for the context! I agree with David about changing the function to `qml.pulse.constant`.

Comment (lillian542) [reply pennylane/pulse/transmon.py]: Since trainability is already a concept in PL, and it's more what @mudit2812 is referring to (trainable input format for parameters), maybe we could instead say something like `# create parametrized frequency with constant magnitude`. I'm not sure if that's the best. But something about making sure this is a parametrized value, rather than trainable per se.

It could also be that it should be in the text beforehand rather than a comment, if it was confusing. Like mention that we want the drive frequency to be an adjustable rather than fixed parameter in the Hamiltonian, so we define it as a callable with a constant output.

I agree with David about switching this out with `freq = qml.pulse.constant` instead.

**Comment by lillian542** [comment]
While we are going back over this, I was looking over the intro, and I would question the accuracy of labelling a and a^dagger the "bosonic creation and annihilation operators" in this context (this is true for the part of the Hamiltonian describing the readout oscillator, but I don't think its true here). We should just call them ladder operators. I would also consider replacing them with the ladder operator symbols sigma^+ and sigma^- to avoid ambiguity between these operators, the bosonic annihilation/creation operators, and the fermionic creation/annihilation operators.

What are others' thoughts on this @dwierichs @Qottmann ?

File: pennylane/pulse/transmon.py (hunk 3 above)
Comment (lillian542) [reply pennylane/pulse/transmon.py]: I just read this again and I don't like "frequency with constant magnitude". I take it back, don't say that. But maybe just `# create parametrized frequency` and some explanatory text above?

Comment (BorjaRequena) [reply pennylane/pulse/transmon.py]: I put all of this together updating the text and the code example. Let me know what you think 😄

**Comment by dwierichs** [comment]
>While we are going back over this, I was looking over the intro, and I would question the accuracy of labelling a and a^dagger the "bosonic creation and annihilation operators" in this context (this is true for the part of the Hamiltonian describing the readout oscillator, but I don't think its true here). We should just call them ladder operators.
I would also consider replacing them with the ladder operator symbols sigma^+ and sigma^- to avoid ambiguity between these operators, the bosonic annihilation/creation operators, and the fermionic creation/annihilation operators.
What are others' thoughts on this

I must admit, I am not sure about the used notational conventions regarding transmon hardware, so I don't have a strong opinion here. Why do you think the name "bosonic creation and annihilation operators" is not apt?

**Comment by lillian542** [comment]
> Why do you think the name "bosonic creation and annihilation operators" is not apt?

So, this came up because I started working on the fermionic operators stuff (where in qchem we've implemented the fermionic creation and annihilation operators).  Having gone over it a bit after confusing myself, I think this is ambiguous at best and incorrect at worst, though I'm not confident which one it is. But here's what I've got.

**What I'm confident about is:** In transom Hamiltonians, you often see a full description that includes the readout resonator. This includes $a$ and $a^+$ operators, and those *are* bosonic creation and annihilation operators, but they describe what is happening with the readout resonator, not the qubit, and they do in fact literally describe the creation or annihilation of photons. For the qubits, you use ladder operators, $\sigma^+$ and $\sigma^-$, to indicate raising or lowering from one qubit state to the next. So at least, this does not fit the convention, where $a$ / $a^+$ is for the readout resonator, and $\sigma^\pm$ for the qubit state. [Here's a pretty standard formulation if you want more details.](https://en.wikipedia.org/wiki/Jaynes–Cummings_model#Mathematical_formulation_1)

**What I'm less confident about, but still believe is correct, is:** when we describe the qubit states, using $a$ vs $\sigma$ is not just a convention to differentiate. These are, in fact, not bosonic creation and annihilation operators. What we mean to say is that we excite from one state to another. To describe this in creation and annihilation operators, we would have to do something like (to go from 0 to 1), $a_0a_1^+$, i.e. remove the excitation on 0 and add one on one. So what we mean by $\sigma^+$ operating on state 0 is *perhaps* the same as $a_0a_1^+$, but is not the same as $a_0^+$, which is what I would argue we are equating it to here. Terminology section of [this wikipedia article](https://en.wikipedia.org/wiki/Ladder_operator) is helpful here. That's not even touching on whether the state-switching we are doing can actually be described as creating and annihilation bosons; perhaps they can, I honestly really don't know.

@dwierichs

File: pennylane/pulse/transmon.py (hunk 3 above)
Comment (Qottmann) [reply pennylane/pulse/transmon.py]: Looks good to me!

**Comment by Qottmann** [comment]
Re bosonic vs ladder operators: @lillian542 I think you are right and those are indeed not bosonic but ladder operators. The derivation of circuit QED comes from quantizing conjugate variables flux and charge, leading to creation and annihilation operators, see e.g. this nice [qiskit introduction to transmon physics](https://learn.qiskit.org/course/quantum-hardware-pulses/introduction-to-transmon-physics). Those are not bosonic (evenly spaced) but anharmonic (unevenly spaced). I somehow thought that by introducing the extra anharmonicity terms this already captures this fact (unevenly spaced energy levels), but reading this again it doesnt seem so. Nice catch!

**Comment by dwierichs** [comment]
> These are, in fact, not bosonic creation and annihilation operators. What we mean to say is that we excite from one state to another. To describe this in creation and annihilation operators, we would have to do something like (to go from 0 to 1),
$a_0a_1^\dagger$, i.e. remove the excitation on 0 and add one on one.

I think here we need to be really clear about what our states are describing: When mixing the perspective of creation/annihilation into two-level systems, I read the states as $|0>$: There is no excitation and $|1>$: There is one excitation. This means that there is not really an operator $a_0$, because the $0$ labels the _absence of any excitation_, rather than the presence of the zero-th mode excitation.
However, this is just the perspective I'm used to, it might not be the correct one here.

**Comment by lillian542** [comment]
> Note though that the readout resonator is typically not taken into account explicitly as we look at effective Hamiltonians of the transmon.

The Jaynes-Cummings model _does_ explicitly contain the readout resonator, and is the formulation I am most familiar with and have seen most. It's derived on page 15 of the circuit QED review you linked (eq 34). To me, this is the most standard formulation of a transom Hamiltonian (but maybe its not an 'effective Hamiltonian'?). We don't take the resonator elements into account, since we are not simulating the resonator, but I think we should avoid 'reassigning' variables that are already meaningful in describing transmons in other common settings. And $a$ is already taken for describing changes in readout resonator state. I'm happy with $b$ and $b^\dagger$.

**Comment by Qottmann** [comment]
Unfortunately there doesnt seem to be a wide consensus for the letters as papers such as e.g. the [ctrl-vqe paper](https://arxiv.org/abs/2008.04302) and [its followup](https://arxiv.org/pdf/2203.06818.pdf) use $a$ and $a^\dagger$ for the superconducting circuit modes. 

Note also that JC model describes the interaction between a bosonic mode and a two-level system, so it already incorporates the approximation of the transmon to only two levels to start with. There are various ways of deriving the effective Hamiltonian, by which I mean the effective description of (just) the qubits (or qudits).

I think what is most important is that we are as explicit as possible to avoid confusion. And I agree that $a$ and $a^\dagger$ is not a good choice for the transmon modes 👍

**Comment by Qottmann** [comment]
Note:
after meeting decided to
- scratch "bosonic" and
- use $b$ and $b^\dagger$ for the creation and annihilation operators

**Review by Qottmann — APPROVED** [review]
looks good to me, thanks @BorjaRequena 👌
